File: module-11-rl-agents/solutions/python/environment.py
"""Module 11: Agent environment."""
import os
import sys
import tempfile
import shutil

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "module-4-filesystem", "solutions", "python"))
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "module-5-coding-agent", "solutions", "python"))
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "module-3-tools", "solutions", "python"))

from sandbox import Sandbox
from tools import create_file_tools
from edit_file import create_edit_file_tool
from tool_registry import ToolRegistry, Tool


class AgentEnvironment:
    def __init__(self):
        self.state = None

    def reset(self, setup: dict) -> dict:
        workspace_dir = tempfile.mkdtemp(prefix="rl-env-")
        sandbox = Sandbox(workspace_dir)

        for file_path, content in setup.get("files", {}).items():
            full_path = os.path.join(workspace_dir, file_path)
            os.makedirs(os.path.dirname(full_path), exist_ok=True)
            with open(full_path, "w") as f:
                f.write(content)

        registry = ToolRegistry()
        for t in create_file_tools(sandbox):
            registry.register(Tool(**t))
        registry.register(Tool(**create_edit_file_tool(sandbox)))

        self.state = {
            "workspace_dir": workspace_dir,
            "sandbox": sandbox,
            "registry": registry,
            "step": 0,
            "max_steps": 20,
            "done": False,
        }
        return self.state

    def step(self, tool_name: str, inp: dict) -> dict:
        if not self.state:
            raise RuntimeError("Environment not initialized. Call reset() first.")
        if self.state["done"]:
            raise RuntimeError("Episode is done. Call reset().")

        self.state["step"] += 1
        result = self.state["registry"].execute(tool_name, inp)

        if self.state["step"] >= self.state["max_steps"]:
            self.state["done"] = True

        return {"result": result, "done": self.state["done"], "step": self.state["step"]}

    def cleanup(self):
        if self.state:
            shutil.rmtree(self.state["workspace_dir"], ignore_errors=True)
            self.state = None

    def get_state(self):
        return self.state

File: module-11-rl-agents/solutions/environment.ts
/**
 * Module 11: Agent environment
 * Sandboxed workspace with tool access and state reset between episodes.
 */
import * as fs from "fs/promises";
import * as path from "path";
import * as os from "os";
import { Sandbox } from "../../module-4-filesystem/solutions/sandbox.js";
import { createFileTools } from "../../module-4-filesystem/solutions/tools.js";
import { createEditFileTool } from "../../module-5-coding-agent/solutions/edit-file.js";
import { ToolRegistry } from "../../module-3-tools/solutions/tool-registry.js";
import type { Tool } from "../../module-2-agent-loop/solutions/types.js";

export interface EnvironmentState {
  workspaceDir: string;
  sandbox: Sandbox;
  registry: ToolRegistry;
  step: number;
  maxSteps: number;
  done: boolean;
}

export interface EpisodeSetup {
  files: Record<string, string>;
  prompt: string;
}

export class AgentEnvironment {
  private state: EnvironmentState | null = null;

  async reset(setup: EpisodeSetup): Promise<EnvironmentState> {
    // Create fresh workspace
    const workspaceDir = await fs.mkdtemp(path.join(os.tmpdir(), "rl-env-"));
    const sandbox = new Sandbox(workspaceDir);

    // Write setup files
    for (const [filePath, content] of Object.entries(setup.files)) {
      const fullPath = path.join(workspaceDir, filePath);
      await fs.mkdir(path.dirname(fullPath), { recursive: true });
      await fs.writeFile(fullPath, content, "utf-8");
    }

    // Build tool registry
    const registry = new ToolRegistry();
    for (const tool of createFileTools(sandbox)) registry.register(tool);
    registry.register(createEditFileTool(sandbox));

    this.state = {
      workspaceDir,
      sandbox,
      registry,
      step: 0,
      maxSteps: 20,
      done: false,
    };

    return this.state;
  }

  async step(toolName: string, input: Record<string, unknown>): Promise<{
    result: string;
    done: boolean;
    step: number;
  }> {
    if (!this.state) throw new Error("Environment not initialized. Call reset() first.");
    if (this.state.done) throw new Error("Episode is done. Call reset().");

    this.state.step++;
    const result = await this.state.registry.execute(toolName, input);

    if (this.state.step >= this.state.maxSteps) {
      this.state.done = true;
    }

    return { result, done: this.state.done, step: this.state.step };
  }

  async cleanup(): Promise<void> {
    if (this.state) {
      await fs.rm(this.state.workspaceDir, { recursive: true, force: true });
      this.state = null;
    }
  }

  getState(): EnvironmentState | null {
    return this.state;
  }
}
